### PR TITLE
GHC convert comment

### PR DESCRIPTION
**Pull request by shayne-fletcher-da**
This PR converts the comment hints to use the GHC parse tree.

(1) `Tests passed (674)`.

(2) `No hints`.

(3) `{- LANGUAGE FlexibleContents #-}`

Screen Shot 2019-06-28 at 11 53 22 AM

(4) `{- LANGUAGE FlexibleContents -}`

Screen Shot 2019-06-28 at 11 54 08 AM

### CODE REVIEW

File: src/HSE/All.hs [hunk 1]
```diff
@@ -183,6 +193,19 @@ data ModuleEx = ModuleEx {
   , ghcAnnotations :: GHC.ApiAnns
 }
 
+-- | The two representations of comments in a tidy little bundle. Used
+-- by 'commentHint'.
+data CommentEx = CommentEx {
```
Comment (ndmitchell) [inline src/HSE/All.hs]: I think we should ditch CommentEx

File: src/GHC/Util.hs [hunk 2]
```diff
@@ -214,3 +218,28 @@ enumerateExtensions = [Cpp .. StarIsType]
 readExtension :: String -> Maybe Extension
 readExtension x = Map.lookup x exts
   where exts = Map.fromList [(show x, x) | x <- [Cpp .. StarIsType]]
+
+trimCommentStart s
+  | "{-" `isPrefixOf` s = fromJust (stripPrefix "{-" s)
```
Comment (ndmitchell) [inline src/GHC/Util.hs]: 4 space indents please

Comment (ndmitchell) [reply src/GHC/Util.hs]:
```suggestion
    | Just s <- stripPrefix "{-" s = s
```

File: src/GHC/Util.hs [hunk 3]
```diff
@@ -214,3 +218,28 @@ enumerateExtensions = [Cpp .. StarIsType]
...
+  | "--" `isPrefixOf` s = fromJust (stripPrefix "--" s)
```
Comment (ndmitchell) [inline src/GHC/Util.hs]: Ditto for stripPrefix here.

File: src/GHC/Util.hs [hunk 4]
```diff
@@ -214,3 +218,28 @@ enumerateExtensions = [Cpp .. StarIsType]
...
+  | otherwise = s
+
+trimCommentEnd s
+  | "-}" `isSuffixOf` s = fromJust (stripSuffix "-}" s)
```
Comment (ndmitchell) [inline src/GHC/Util.hs]: Ditto with a pattern guard.

File: src/HSE/All.hs [hunk 5]
```diff
@@ -183,6 +193,19 @@ data ModuleEx = ModuleEx {
...
+    hseComment :: Comment
+  , ghcComment :: Located GHC.AnnotationComment
+}
+
+-- | Extract a list of all a parsed module's comments.
+comments :: ModuleEx -> [CommentEx]
```
Comment (ndmitchell) [inline src/HSE/All.hs]: Can we get ghcComments promoted to a top-level thing, as I imagine it's going to be useful.

Comment (ndmitchell) [reply src/HSE/All.hs]: And kill comments of course, since there will be no CommentEx

File: src/Hint/Type.hs [hunk 6]
```diff
@@ -23,7 +23,7 @@ data Hint {- PUBLIC -} = Hint
     , hintDecl :: [Setting] -> Scope -> ModuleEx -> Decl SrcSpanInfo -> [Idea]
         -- ^ Given a declaration (with a module and scope) generate some 'Idea's.
         --   This function will be partially applied with one module/scope, then used on multiple 'Decl' values.
-    , hintComment :: [Setting] -> Comment -> [Idea] -- ^ Given a comment generate some 'Idea's.
+    , hintComment :: [Setting] -> CommentEx -> [Idea] -- ^ Given a comment generate some 'Idea's.
```
Comment (ndmitchell) [inline src/Hint/Type.hs]: Let's kill hintComment entirely

File: src/Config/Haskell.hs [hunk 7]
```diff
@@ -78,22 +81,23 @@ readPragma o = case o of
         f _ _ = Nothing
 
 
-readComment :: Comment -> [Classify]
-readComment o@(Comment True _ x)
+readComment :: CommentEx -> [Classify]
+readComment c@CommentEx {ghcComment=L pos (AnnBlockComment s)}
```
Comment (ndmitchell) [inline src/Config/Haskell.hs]: This will become `c@(L pos AnnBlockComment{})` in the new world.

File: src/GHC/Util.hs (hunk 2 above)
Comment (shayne-fletcher-da) [reply src/GHC/Util.hs]: ✔️

File: src/GHC/Util.hs (hunk 4 above)
Comment (shayne-fletcher-da) [reply src/GHC/Util.hs]: ✔️

File: src/GHC/Util.hs (hunk 3 above)
Comment (shayne-fletcher-da) [reply src/GHC/Util.hs]: ✔️

File: src/HSE/All.hs (hunk 1 above)
Comment (shayne-fletcher-da) [reply src/HSE/All.hs]: ✔️

File: src/HSE/All.hs (hunk 5 above)
Comment (shayne-fletcher-da) [reply src/HSE/All.hs]: ✔️

File: src/Config/Haskell.hs (hunk 7 above)
Comment (shayne-fletcher-da) [reply src/Config/Haskell.hs]: ✔️

File: src/Hint/Type.hs (hunk 6 above)
Comment (shayne-fletcher-da) [reply src/Hint/Type.hs]: ✔️

**Comment by shayne-fletcher-da** [comment]
@ndmitchell , if you'd be so kind, I think this is ready for another look.

**Comment by ndmitchell** [comment]
Looks awesome, thanks

**Pull request merged**